### ISSUE TITLE
datastore: Panics when confronted with an absent credentials file

### ISSUE THREAD

**Issue by Freddixx** [googleapis/google-cloud-go#3236]
Hi all 👋 

this probably sits in between a bug and a feature request, but I decided to file it as a bug. Feel free to move it.

Our service panics when the datastore client is configured with a non-existent credentials file. 

More details below, let me know if you have any questions! 🙂 

**Client**
Datastore lib - go client

**Environment**
Really all I believe

**Go Environment**

$ go version
`> go1.14.2 darwin/amd64`
$ go env
```
GO111MODULE=""
GOARCH="amd64"
GOBIN=""
GOCACHE="/Users/XXXXXXXX/Library/Caches/go-build"
GOENV="/Users/XXXXXXX/Library/Application Support/go/env"
GOEXE=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="darwin"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB=""
GOOS="darwin"
GOPATH="/Users/XXXXXX/go"
GOPRIVATE=""
GOPROXY="https://proxy.golang.org,direct"
GOROOT="/usr/local/Cellar/go/1.14.2_1/libexec"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/Cellar/go/1.14.2_1/libexec/pkg/tool/darwin_amd64"
GCCGO="gccgo"
AR="ar"
CC="clang"
CXX="clang++"
CGO_ENABLED="1"
GOMOD="XXXX"
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fno-caret-diagnostics -Qunused-arguments -fmessage-length=0 -fdebug-prefix-map=/var/folders/wl/gtxdxzn54g352n1fzzsvdvf00000gn/T/go-build788020205=/tmp/go-build -gno-record-gcc-switches -fno-common"

```

**Code**
```go
client, err := datastore.NewClient(
	context,
	projectID,
	option.WithCredentialsFile(config.Get("credentials")),
)

...
// Context:
// browserConfigurationKind = "BrowserConfiguration"
// query = datastore.NewQuery(browserConfigurationKind).Namespace(config.Get("datastore_namespace"))
var browsers []models.BrowserConfiguration
keys, err := datastoreService.client.GetAll(datastoreService.context, query, &browsers)
```

**Expected behavior**
A readable error that the provided credentials could not be processed. This can then be handled and logged.

**Actual behavior**
Datastore panics and produces the following error:

```
2020/11/19 13:50:22 http: panic serving 127.0.0.1:52862: runtime error: invalid memory address or nil pointer dereference
goroutine 62 [running]:
net/http.(*conn).serve.func1(0xc0006b0460)
	/usr/local/go/src/net/http/server.go:1801 +0x147
panic(0xb8bc20, 0x11add90)
	/usr/local/go/src/runtime/panic.go:975 +0x47a
cloud.google.com/go/datastore.(*Client).Run(0x0, 0xd5c5c0, 0xc0006af140, 0xc0006d20f0, 0xd69120)
	/go/pkg/mod/cloud.google.com/go/datastore@v1.3.0/query.go:580 +0x164
cloud.google.com/go/datastore.(*Client).GetAll(0x0, 0xd5c540, 0xc000038058, 0xc0006d20f0, 0xb1eae0, 0xc0003f62c0, 0x0, 0x0, 0x0, 0x0, ...)
	/go/pkg/mod/cloud.google.com/go/datastore@v1.3.0/query.go:520 +0x157
github.com/XXX/whitelist/services/datastore.(*datastoreService).getAllBrowserConfigurationsWithQuery(0x11c67b0, 0xc0006d20f0, 0xc00001e000, 0xb98660, 0xb986e0)
	/app/services/datastore/datastore.go:24 +0x8d
github.com/XXX/whitelist/services/datastore.(*datastoreService).GetBrowserConfigurations(0x11c67b0, 0x1, 0x0, 0xc000107970)
	/app/services/datastore/datastore.go:44 +0x3c
github.com/XXX/whitelist/api/browser-whitelist.getBrowsersWhitelist(0xd5a4c0, 0xc0006c8540, 0xc0006bfe00)
	/app/api/browser-whitelist/browser_whitelist.go:29 +0x2d
net/http.HandlerFunc.ServeHTTP(0xca29c0, 0xd5a4c0, 0xc0006c8540, 0xc0006bfe00)
	/usr/local/go/src/net/http/server.go:2042 +0x44
goji%2eio.dispatch.ServeHTTP(0xd5a4c0, 0xc0006c8540, 0xc0006bfe00)
	/go/pkg/mod/goji.io@v2.0.2+incompatible/dispatch.go:17 +0xba
goji%2eio.(*Mux).ServeHTTP(0xc0008a6700, 0xd5a4c0, 0xc0006c8540, 0xc0006bfc00)
	/go/pkg/mod/goji.io@v2.0.2+incompatible/mux.go:74 +0x1d1
goji%2eio.dispatch.ServeHTTP(0xd5a4c0, 0xc0006c8540, 0xc0006bfc00)
	/go/pkg/mod/goji.io@v2.0.2+incompatible/dispatch.go:17 +0xba
github.com/XXX/whitelist/middleware/requesthandler.SetHeaders.func1(0xd5a4c0, 0xc0006c8540, 0xc0006bfc00)
	/app/middleware/requesthandler/request_handler.go:17 +0x308
net/http.HandlerFunc.ServeHTTP(0xc000094ee0, 0xd5a4c0, 0xc0006c8540, 0xc0006bfc00)
	/usr/local/go/src/net/http/server.go:2042 +0x44
goji%2eio.(*Mux).ServeHTTP(0xc0008a6680, 0xd5a4c0, 0xc0006c8540, 0xc0006bf900)
	/go/pkg/mod/goji.io@v2.0.2+incompatible/mux.go:74 +0x1d1
net/http.serverHandler.ServeHTTP(0xc0008e21c0, 0xd5a4c0, 0xc0006c8540, 0xc0006bf900)
	/usr/local/go/src/net/http/server.go:2843 +0xa3
net/http.(*conn).serve(0xc0006b0460, 0xd5c500, 0xc00003d700)
	/usr/local/go/src/net/http/server.go:1925 +0x8ad
created by net/http.(*Server).Serve

```

**Additional context**
We're using a vault to inject secrets dynamically into our containers. Due to a misconfiguration the injected credentials file was absent.

**Comment by crwilcox** [comment]
@Freddixx I was hoping to get a little clarity here. 

`config.Get("credentials")` what is being returned here? Is this nil, or something else?

**Comment by Freddixx** [comment]
@crwilcox, I will try to repro the behaviour on a running container and dump the contents of that variable.

I'll ping you back as soon as I did that.

**Comment by Freddixx** [comment]
@crwilcox I was trying to reproduce the exact same environment in the past two days. I got very close, but this time I got an error message upon connecting to datastore:

```Failed to initialize the datastore client dialing: cannot read credentials file: open /secrets/v1/datastore: no such file or directory"```

I am 100% sure that message didn't exist last time (hence this issue).

Anyhow, this _does qualify_ as a readable error message and hence the lib does what it's supposed to. 

I will close this issue for the time being and report back with more info if the situation ever comes up again.

Sorry for the troubles!
